Support custom product attributes in advanced filter

jeffstieler opened the pull request:
Fixes #5560, along with https://github.com/woocommerce/woocommerce-admin/pull/5840.

This PR hooks up the product attribute advanced filter to the new endpoints that support custom attributes.

To support that end:
* A new `attributes` autocompleter was added to the `<Search>` component
* The `<Search>` component `selected` prop was updated to match the inner `<SelectControl>` that it gets passed to
* The passthough `multiple` prop can be defined on the `<Search>` component
* The attribute filter now uses the "inline tag" style for chosen values
* Some UX tweaks were made for `<SelectControl>` for combinations of `multiple` and `inlineTags`
* The reports datastore `get_attribute_subqueries()` method had custom attribute support added

### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

-   [x] I've tested using only a keyboard (no mouse)
-   [x] I've tested using a screen reader
-   [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
-   [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

![Screen Shot 2020-12-09 at 9 55 48 AM](https://user-images.githubusercontent.com/63922/101660918-b0363980-3a15-11eb-9ffc-53f05f5e9026.png)

Searching returns both custom and global product attributes.

![Screen Shot 2020-12-09 at 9 56 43 AM](https://user-images.githubusercontent.com/63922/101660968-bc21fb80-3a15-11eb-82ab-ccd02a49b9fa.png)

Selecting a custom attribute.

![Screen Shot 2020-12-09 at 9 56 57 AM](https://user-images.githubusercontent.com/63922/101661015-c80dbd80-3a15-11eb-9af9-226ad2710b6b.png)


### Detailed test instructions:

- Create an order using Woo's sample product data variable Hoodie /product/hoodie/?attribute_pa_color=green&attribute_logo=No.
- Confirm the Hoodie has variations "Logo" and "Color" in the Edit Product screen.
- See the data from the purchase available in the Orders Report.
- Advanced Filters > Add a filter > Attribute
- See the "logo" attribute available in the dropdown
- Select "no" for the value
- Click "Filter"
- Verify the order is in the results
- Refresh the page
- Verify the filter values are restored from the URL querystring

<!--- Please add a Changelog note

Enter a changelog note following the WooCommerce core format using prefixes of Enhancement:, Tweak:, Dev:, Fix:, Performance: to readme.txt under the "unreleased" list at the top of the changelog. If none exists, please add it. Also include PR number.

If changes pertain to a package, also update CHANGELOG.md in the package's folder in a similar manner.--->

jeffstieler committed:
Add product attributes autocomplete for Search control.

jeffstieler committed:
Support single selection in Search component.

jeffstieler committed:
Fix SelectControl input after selection in single mode.

jeffstieler committed:
Remove label from Search component's inner SelectControl.

Placeholder is the behavior we want to avoid label and selection collision.

jeffstieler committed:
Match Search component selected prop to SelectControl.

It's passed through anyhow.

jeffstieler committed:
Rework attribute filter to use new endpoints for custom terms.

jeffstieler committed:
Hook up custom attribute filter to report queries.

jeffstieler committed:
Fix errant space in filter name.

jeffstieler committed:
Fix SelectControl active status when control is disabled.

jeffstieler committed:
Fix SelectControl "has tags" logic for non-multiple use cases.

jeffstieler committed:
Fix attribute filter screen reader text.

jeffstieler committed:
Update changelogs.

becdetat commented on packages/components/src/select-control/control.js:
```diff
@@ -122,8 +122,11 @@ class Control extends Component {
+			selected,
 		} = this.props;
 		const { isActive } = this.state;
+		const disabled = ! multiple && inlineTags && selected.length > 0;
```
You could include `isActive` in this calculation and simplify the logic further on

jeffstieler replied:
I've removed this code.. I found it to be problematic (keeping `isActive` state but disabling the nested `input`, etc etc) - the UX without disabling the nested `input` isn't all that confusing anyhow.

becdetat commented on src/API/Reports/DataStore.php:
```diff
@@ -1322,7 +1330,7 @@ protected function get_filtered_ids( $query_args, $field, $separator = ',' ) {
 		 * @param string $context    The data store context.
 		 */
-		$ids = apply_filters( 'woocommerce_analytics_ ' . $field, $ids, $query_args, $field, $this->context );
+		$ids = apply_filters( 'woocommerce_analytics_' . $field, $ids, $query_args, $field, $this->context );
```
nice catch

becdetat commented on tests/api/reports-orders.php:
```diff
@@ -231,6 +248,22 @@ public function test_product_attributes_filter() {
 		$this->assertEquals( 1, count( $response_orders ) );
 		$this->assertEquals( $response_orders[0]['order_id'], $variation_order_2->get_id() );
+
+		// Filter by the "Numeric Size" custom attribute, with value "1".
```
Can this new test be added to a new test function?

becdetat reviewed (approved):
Nice! Works well and the code looks good apart from a couple of minor optional suggestions.

jeffstieler committed:
Move custom attribute test to it's own function.

jeffstieler committed:
Remove SelectControl input disabling.

This was supposed to improve keyboard behavior for non-multiple controls, but it was questionable.

jeffstieler committed:
Fix backspace in attribute term field.

jeffstieler commented:
Thanks for the review @becdetat! Since I've made some tweaks, I'd really appreciate another look. 🙇

becdetat commented:
@jeffstieler Looks good! It still all works well. Thanks for the changes.

AlexPl81 commented:
Hello, i'm still having an issue with Analytics>Order using Advanced Filter that Attribute Values are limited to first 10 alphabetically chosen ones. No ability to search either. Can't see any errors in the console. I can see that this issue is supposed to be solved, but it is present for me for some reason.
